### PR TITLE
Dyadic and extended arithmetics

### LINKED ISSUE #9: Add Approx constants and rounding
## Details:
Hours Spent: 0

## Background
Part of epic #8.

## Overview
Add tools to create `Approx` objects for integer and rational numbers, rounded with a given accuracy.

## Acceptance Criteria
- [x] function for rounding an `Approx` to a given mBound (in PR #21)
- [x] function for comparing two `Approx' for set inclusion
- [x] functions for approximating a given integer (or big integer) as an `Approx`
- [x] function for approximating a rational number as an `Approx`

## Closing Notes

Overview of changes can be found in PR #45

### LINKED ISSUE #20: Add Approx field operations
## Details:
Hours Spent: 0

## Background
This is part of epic #8. Field operations (+, -, *, /) are essential for computing function plots.

## Overview
Port the operations from relevant parts of CDAR to form a Field type class instance.

## Acceptance Criteria
- [x] instance Semiring Approx
- [x] instance Ring Approx
- [x] instance EuclideanRing Approx
- [x] instance Field Approx

## Closing Notes

Overview of changes can be found in PR #45

### PR DESCRIPTION
**Pull request by michalkonecny**
# Issues
- Working towards #9 and #20, but not closing them yet

# Summary of changes
- Added Extended wrapper for numerical types, adding positive and negative infinity to the type, while preserving properties of a commutative ring
- Added conversion from Dyadic to Rational
- Added a generic multiplicative power operator

# Unit test changes
- Added tests that Dyadic numbers form a commutative ring and are totally ordered
- Added test of conversion from Dyadic to Rational commutes with addition
- Added facilities for testing that any numerical type is a commutative ring or a total order, giving useful messages when errors are found

### CODE REVIEW

File: src/IntervalArith/Extended.purs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  mul a NegInf = if a < zero then PosInf else NegInf
+  mul NegInf a = if a < zero then PosInf else NegInf
+  mul a b = (*) <$> a <*> b
+
+instance ringExtended :: (Ord a, Ring a) => Ring (Extended a) where
+  sub = lift2 sub
+
+instance commutativeringExtended :: (Ord a, CommutativeRing a) => CommutativeRing (Extended a)
+
+--     abs = fmap abs
```
Comment (RelativeForce) [inline src/IntervalArith/Extended.purs]: Removed zombie code

**Review by RelativeForce — CHANGES_REQUESTED** [review]

File: src/IntervalArith/Extended.purs (hunk 1 above)
Comment (michalkonecny) [reply src/IntervalArith/Extended.purs]: Thank you for looking into this.  I am confused by this request though.  Firstly, GitHub shows four lines being added but I suspect that it is a confusing way of telling me that the four lines are being deleted.  I will assume this is the case. 

I do not think that the four lines should be deleted. The CommutativeRing instance is important, even though this class does not add any operations.  It is there to declare that the multiplication for this type is commutative.  

The commented out definition of abs is part of what I copied from the Haskell version ready for conversion to PureScript when we need it.  I guess this and the other 7 commented out lines could be deleted and brought back from the Haskell version again if you prefer.  What do you think?

Comment (RelativeForce) [reply src/IntervalArith/Extended.purs]: It is not good practice to have commented out code in a codebase. The main reason is that when refactoring or new work is undertaken, the "zombie" code has to be refactored as well which is a waste time as it is not used. That or it becomes outdated and no longer works when it is needed.

If your 100% sure we will need it then I'll approve the PR

Comment (michalkonecny) [reply src/IntervalArith/Extended.purs]: Makes sense.  I have removed/finished the Zombie code.  Thank you!

**Review by RelativeForce — APPROVED** [review]
Looks good

**Pull request merged**